test/common/router/router_test.cc

```diff
@@ -4908,6 +4908,30 @@ TEST_F(RouterTest, PropagatesUpstreamFilterState) {
       "upstream data"));
 }
 
+TEST_F(RouterTest, PropagatesShadowState) {
+  NiceMock<Http::MockRequestEncoder> encoder;
+  Http::ResponseDecoder* response_decoder = nullptr;
+  ON_CALL(callbacks_.stream_info_, isShadow()).WillByDefault(Return(true));
+
+  // This pattern helps ensure that we're actually invoking the callback.
+  bool shadow_state_verified = false;
+  router_->config().upstream_logs_.push_back(std::make_shared<TestAccessLog>(
+      [&](const auto& stream_info) { shadow_state_verified = stream_info.isShadow(); }));
+  expectResponseTimerCreate();
+  expectNewStreamWithImmediateEncoder(encoder, &response_decoder, Http::Protocol::Http10);
+
+  Http::TestRequestHeaderMapImpl headers{};
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_->decodeHeaders(headers, true);
+
+  Http::ResponseHeaderMapPtr response_headers(
+      new Http::TestResponseHeaderMapImpl{{":status", "200"}});
+  // NOLINTNEXTLINE: Silence null pointer access warning
```

Comment:
I think you could `ASSERT_THAT(response_decoder, testing::IsNotNull());` after it's populated, and avoid the need for nolint.

If that doesn't work, `ASSERT(response_decoder != nullptr);` might.

Reply:
Done, thanks for the suggestion. I don't want to pollute blame layer by doing the same elsewhere in this file.